### PR TITLE
Prepare rust rules for Starlark CcToolchainInfo.

### PR DESCRIPTION
**Pull request by buildbreaker2021**
https://github.com/bazelbuild/rules_rust/issues/2425

### CODE REVIEW

File: rust/private/rustc.bzl [hunk 1]
```diff
@@ -671,7 +671,12 @@ def collect_inputs(
     """
     linker_script = getattr(file, "linker_script") if hasattr(file, "linker_script") else None
 
-    linker_depset = cc_toolchain.linker_files()
+    # TODO(gnish): rules_rust is not coupled with Bazel release. Remove conditional and change to
```
Comment (UebelAndre) [inline rust/private/rustc.bzl:674]: Can you instead reference a public github issue or something? The Google `TODO(name)` comments are not that useful for external folks.

**Review by UebelAndre — CHANGES_REQUESTED** [review]

File: rust/private/rustc.bzl (hunk 1 above)
Comment (buildbreaker2021) [reply rust/private/rustc.bzl:674]: Unfortunately there is no public github issue - I will remove TODO. I think the comment is useful even without an issue it clearly explains why this change is necessary and when it should be removed. I can extend it if you'd like to.

Comment (buildbreaker2021) [reply rust/private/rustc.bzl]: I removed username, and updated the comment to better describe the issue.

**Comment by buildbreaker2021** [comment]
Friendly ping

File: rust/private/rustc.bzl [hunk 2]
```diff
@@ -671,7 +671,14 @@ def collect_inputs(
     """
     linker_script = getattr(file, "linker_script") if hasattr(file, "linker_script") else None
 
-    linker_depset = cc_toolchain.linker_files()
+    # TODO: As of writing this comment Bazel used Java CcToolchainInfo.
```
Comment (daivinhtran) [inline rust/private/rustc.bzl:674]: Do you mind filing a Github issue and put it here so we can triage later? It's hard to keep track of all the TODOs.

Comment (buildbreaker2021) [reply rust/private/rustc.bzl:674]: Done.

File: rust/private/rustc.bzl (hunk 1 above)
Comment (buildbreaker2021) [reply rust/private/rustc.bzl]: Done.

**Review by UebelAndre — APPROVED** [review]
The underscore naming convention makes me nervous that we’re setting ourselves up to need to do some crazy migration and would rather not start relying on deprecated/private interfaces in the first place. This change seems fine to me but am more concerned about the conversion process for rules_cc. I also think there should be an issue made there that tracks this effort

**Comment by buildbreaker2021** [comment]
> The underscore naming convention makes me nervous that we’re setting ourselves up to need to do some crazy migration and would rather not start relying on deprecated/private interfaces in the first place.

About private part - API used by `rules_rust` was private in the first place. `rules_rust` is allowlisted so it can access Bazel's internal private API.

About migration part - I *think* the most likely scenario here is that we make that parameter public eventually. Most likely the migration which will need to be done would be to migrate from `_linker_files` to `linker_files`.

**Pull request merged**